test/base/src/androidTest/kotlin/io/realm/kotlin/test/shared/RealmSetTests.kt

```diff
@@ -0,0 +1,1087 @@
+                findLatest(container)!!.objectSetField.add(RealmSetContainer())
+            }
+        }
+    }
+
+    private fun getCloseableRealm(): Realm =
+        RealmConfiguration.Builder(schema = setOf(RealmSetContainer::class))
+            .directory(tmpDir)
+            .name("closeable.realm")
+            .build()
+            .let {
+                Realm.open(it)
+            }
+
+    private fun getTypeSafety(classifier: KClassifier, nullable: Boolean): SetTypeSafetyManager<*> =
+        when {
+            nullable -> NullableSet(
+                property = RealmSetContainer.nullableProperties[classifier]!!,
+                dataSetToLoad = getDataSetForClassifier(classifier, true)
+            )
+            else -> NonNullableSet(
+                property = RealmSetContainer.nonNullableProperties[classifier]!!,
+                dataSetToLoad = getDataSetForClassifier(classifier, false)
+            )
+        }
+
+    @Suppress("UNCHECKED_CAST", "ComplexMethod")
+    private fun <T> getDataSetForClassifier(
+        classifier: KClassifier,
+        nullable: Boolean
+    ): List<T> = when (classifier) {
+        Byte::class -> if (nullable) NULLABLE_BYTE_VALUES else BYTE_VALUES
+        Char::class -> if (nullable) NULLABLE_CHAR_VALUES else CHAR_VALUES
+        Short::class -> if (nullable) NULLABLE_SHORT_VALUES else SHORT_VALUES
+        Int::class -> if (nullable) NULLABLE_INT_VALUES else INT_VALUES
+        Long::class -> if (nullable) NULLABLE_LONG_VALUES else LONG_VALUES
+        Boolean::class -> if (nullable) NULLABLE_BOOLEAN_VALUES else BOOLEAN_VALUES
+        Float::class -> if (nullable) NULLABLE_FLOAT_VALUES else FLOAT_VALUES
+        Double::class -> if (nullable) NULLABLE_DOUBLE_VALUES else DOUBLE_VALUES
+        String::class -> if (nullable) NULLABLE_STRING_VALUES else STRING_VALUES
+        RealmInstant::class -> if (nullable) NULLABLE_TIMESTAMP_VALUES else TIMESTAMP_VALUES
+        ObjectId::class -> if (nullable) NULLABLE_OBJECT_ID_VALUES else OBJECT_ID_VALUES
+        ByteArray::class -> if (nullable) NULLABLE_BINARY_VALUES else BINARY_VALUES
+        RealmObject::class -> SET_OBJECT_VALUES // Don't use the one from RealmListTests!!!
+        else -> throw IllegalArgumentException("Wrong classifier: '$classifier'")
+    } as List<T>
+}
+
+/**
+ * Tester interface defining the operations that have to be tested exhaustively.
+ */
+internal interface SetApiTester<T, Container> {
+
+    val realm: Realm
+
+    override fun toString(): String
+    fun copyToRealm()
+    fun add()
+    fun clear()
+    fun contains()
+    fun iterator()
+    fun iterator_hasNext()
+    fun iterator_next()
+    fun iterator_remove()
+    fun iteratorFailsIfRealmClosed(realm: Realm)
+
+    /**
+     * Asserts structural equality for two given collections. This is needed to evaluate equality
+     * contents of ByteArrays and RealmObjects.
+     */
+    fun assertStructuralEquality(expected: Collection<T>, actual: Collection<T>)
```

Comment:
These methods seem to be very generic in nature, should the live outside this class as free floating functions instead?